src/bin/add/args.rs

```diff
@@ -141,13 +141,42 @@ pub struct Args {
     /// Registry to use
     #[structopt(long = "registry", conflicts_with = "git", conflicts_with = "path")]
     pub registry: Option<String>,
+
+    /// print available features inline. Useful for testing
+    #[structopt(long = "inline")]
+    pub inline: bool,
 }
 
 fn parse_version_req(s: &str) -> Result<&str> {
     semver::VersionReq::parse(s).chain_err(|| "Invalid dependency version requirement")?;
     Ok(s)
 }
 
+pub fn get_featues_from_manifest_path(manifest: &Path) -> Vec<String> {
+    let cargo_path = if manifest.ends_with("Cargo.toml") {
+        manifest.into()
+    } else {
+        manifest.join("Cargo.toml")
+    };
+    //TODO: be better about error handling
+    let dep_package = MetadataCommand::new()
+        .no_deps()
+        .manifest_path(&cargo_path)
+        .exec()
+        .ok()
+        .and_then(|m| m.packages.get(0).cloned());
+
+    match dep_package {
+        Some(pkg) => {
+            let mut tmp = pkg.features.keys().cloned().collect::<Vec<String>>();
+            tmp.sort();
+            tmp.dedup();
+            tmp
+        }
+        None => vec![],
+    }
+}
```

Comment:
Check out https://github.com/killercup/cargo-edit/pull/541.  Once that is merged, it should make some of the work in this PR a little easier.

Reply:
Now that #541 is merged, we should be able to add an `fn features()` to `Manifest`.  We can then have `crate_name` populate it as well as the direct git / path code. 

This makes it so we can later add any extra logic we want to it, like handling aliasing, to `Manifest`.

With all of this, that just leaves the name/version case left to handle, whether in this PR or as a follow up.

Reply:
it indeed cleaned up the logic. I even added the functionality for versioned crates that also have a path. For simply plain versioned crates (e.g. `docopt@0.8`) I didn't implement it just yet because that would require either duplicating or rewriting the logic for fetching from the registry and that felt a bit too much to add to this PR. I imagine that if you're adding specific versions of a crate you are also already aware of the available features and which one you need, but maybe that's just me. If there is a need for that last case we can work on it.